pil_analyzer/src/lib.rs

```diff
@@ -183,16 +184,40 @@ pub enum Expression {
     MatchExpression(Box<Expression>, Vec<(Option<FieldElement>, Expression)>),
 }
 
-#[derive(Debug, PartialEq, Eq, Default, Clone)]
+#[derive(Debug, Clone, Eq)]
 pub struct PolynomialReference {
-    // TODO would be better to use numeric IDs instead of names,
-    // but the IDs as they are overlap. Maybe we can change that.
+    /// Name of the polynomial - just for informational purposes.
+    /// Comparisons are based on polynomial ID.
     pub name: String,
+    /// Identifier for a polynomial reference.
+    /// Optional because it is filled in in a second stage of analysis.
+    /// TODO make this non-optional
+    pub poly_id: Option<(u64, PolynomialType)>,
     pub index: Option<u64>,
```

Comment:
So this one becomes `None` after `poly_id` is filled? If so, would be good to document that here.

Reply:
If it's always one of the other, would it make sense to have:

```
struct PolynomialReference<Id> {
   ..
   index: Id
}
```

And go from `PolynomialReference<u64>` to `PolynomialReference<(u64, PolynomialType)>` in the second stage of analysis? That would remove a lot of the runtime assumptions introduced in this PR (at the cost of making the AST generic over `Id` but maybe that's okay?)

Reply:
But yes, I have considered making it generic, where the first-stage structure would have `poly_id: ()`, and the second-stage would be `poly_id: (u64, PolynomialType)`. Created issue #232  for that.